Support Wagtail 6 + add sharing link to Action Menu + more

stevejalim opened the pull request:
This PR is based on our fork of the project and I'm very happy to offer our changes back upstream.

The changeset:

* adds Wagtail 6 / Stimulus JS support to wagtaildraftsharing
* adds an additional way to generate a sharing link, from the Action Menu at the foot of the page - if there is an unpublished Draft only
* makes some of the wording around 'draftsharing' customisable, in a way that won't trigger migrations if a project does use non-default wording
* expands test coverage, bringing things up to ~98% and effectively 100%
* adds CI using tox and Github Actions 
* sketches in a Release GHA which can be uncommented for use on this upstream repo
* crops Python 3.8 support (which is now EOL), adds 3.9, 3.10, 3.11 and 3.12 to tox. 

For more details, please see the PR on our side: https://github.com/mozmeao/wagtaildraftsharing/pull/1/

----

Note that this hasn't been tested in production yet, so feel free to wait a few weeks before merging. Equally, if during review you spot something that could be done better, please feel free to mention it!

stevejalim committed:
Make draftsharing menu position customisable

stevejalim committed:
Change menu icon

stevejalim committed:
Make draftsharing verbose names customisable

stevejalim committed:
Support current draftsharing behaviour in Wagtail 6

This commit mirrors the current functionality for Wagtail 5 in Wagtail 6 by
creating a Stimulus JS controller and hooking it into Wagtail in a similar
way as for Wagtail 5.

For the page history view, the HTML templating is different, so the previous
approach overrode wagtailadmin/pages/revisions/_actions.html to slot in custom
markup that the JS could hook into. For Wagtail 6 we do something similar,
extending wagtailadmin/generic/history/action_cell.html to slot in a bit of
custom markup.

For snippet-list view, we update the HTML output for the share_url method
to hook in the Stimulus controller used above.

The JS in the Stimulus controller is deliberately very similar to the
original JS. I wondered about refactoring out the main duplicated code
(the main function which triggers the creation of new links via a POST
to the backend, and the function which upgrades the View button on the
list view of existing sharing links to be a clipboard link). However,
for now - and with a view potentially to deprecating pre-Wagtail 6 behaviour
- I've gone with some repetition rather than DRY code.

Note that we don't load both JS files into the browser - we only load
the appropriate one based on the version of Wagtail in use

stevejalim committed:
Add an option in the main Page menu to create a draftsharing link

...if there is an unpublished draft that can be shared

stevejalim committed:
Add missing migration for change to 'editable' attr of created_by field

stevejalim committed:
Update README with new screenshot + make links relative

stevejalim committed:
Expand test coverage for SharingLinkView

stevejalim committed:
Expand test coverage around expiry of links and setting expiry time during link creation

stevejalim committed:
Move link expiry setting into app-level settings

stevejalim committed:
Expand test coverage: access to view to create a sharing link

stevejalim committed:
Expand view test to 100% coverage

stevejalim committed:
Refactor view tests to split into focused TestCases now that we have more tests

stevejalim committed:
Rename snippet class to improve code comprehension a tiny bit

Bit of a nitpick here, I know - sorry!

stevejalim committed:
Add tests for model methods

stevejalim committed:
Naming and formatting tweaks

incl datetime_now -> timezone_now to avoid mis-assumption about which library it's from

stevejalim committed:
Add comments to help understanding

stevejalim committed:
Expand model test coverage to 100%; fixup test regression due to code tidying earlier

stevejalim committed:
Add tests for wagtail_hooks

Brings coverage up to 98%, and effectively all code that needs coverage

stevejalim committed:
Update gitignore to skip build artifacts and coverage reports

stevejalim committed:
Address outlier case where, with Wagtail 5.1, the active_until field was coming back as a naive datetime

So we patch it in the tests and warn about it

stevejalim committed:
Update tox to cover more versions of Python, Wagtail and Django

Note the tolerance of a should-not-build version python3.12-django3.2-wagtail6.1 - may be there are better ways around that

stevejalim committed:
Fix unused import

stevejalim committed:
Switch to only using ruff, not black + isort + ruff

No need to have three things going when ruff does the lot

stevejalim committed:
Formatting fixups with ruff (only)

stevejalim committed:
Add GHA for CI; drop 3.8 support

stevejalim committed:
Remove redundant isort config

stevejalim committed:
Stabilise tests across Python 3.9 and/or Wagtail 5.1 by enforcing UTC expiry timestamps

Note the addition to the README, requiring the use of settings.USE_TZ=True

stevejalim committed:
Update ruff config with known sort order

stevejalim committed:
Merge pull request #1 from mozmeao/add-wagtail-6-support

Add Wagtail 6 support + extra way to create a sharing link

stevejalim commented:
@KIRA009 @zerolab FYI

stevejalim committed:
Fix blow-up when adding a new page to the CMS

stevejalim committed:
Update pyproject.toml with new (mozmeao) version number and add self as a maintainer contact

stevejalim committed:
Merge pull request #2 from mozmeao/15291--hotfix--page-creation-blocked

Hotfix: page creation was broken by addition to Action Menu 🤦

KIRA009 commented on wagtaildraftsharing/settings.py:
```diff
@@ -0,0 +1,32 @@
+
+WAGTAIL_DRAFTSHARING_ADMIN_MENU_POSITION = getattr(
+    settings,
+    "WAGTAIL_DRAFTSHARING_ADMIN_MENU_POSITION",
```
Suggestion: Should we settle on either of `WAGTAIL_DRAFTSHARING` and `WAGTAILDRAFTSHARING` (used for `WAGTAIL_DRAFTSHARING_MAX_AGE`) for uniformity? I have a preference for `WAGTAILDRAFTSHARING` just because it's the 'original', but if we go for `WAGTAIL_DRAFTSHARING`, can we do something like

```python
WAGTAIL_DRAFTSHARING_MAX_AGE = getattr(
    settings,
    "WAGTAIL_DRAFTSHARING_MAX_AGE",
    getattr(
        settings,
        "WAGTAILDRAFTSHARING_MAX_AGE",
        DEFAULT_MAX_AGE,
    )
)
```

and maybe add a system check to add a warning if the user is using the old `WAGTAILDRAFTSHARING_MAX_AGE`, with a message to change it to `WAGTAIL_DRAFTSHARING_MAX_AGE`, before the next release (major?).

stevejalim replied:
Yeah i'm happy to go with `WAGTAILDRAFTSHARING_` - thanks for the sharp eye, and sorry for missing that pattern in the existing setting. 

I've made a PR against our fork, which will then update this PR once it's merged. Let me know what you think via a comment on that PR https://github.com/mozmeao/wagtaildraftsharing/pull/3

KIRA009 reviewed (commented):
Hi @stevejalim I have just one outstanding question regarding the naming of the settings. Thanks for all the hard work, the tests are really thorough

stevejalim committed:
Standardise settings prefix to be WAGTAILDRAFTSHARING_* and update README/docs

stevejalim committed:
Bump mozmeao version

stevejalim committed:
Merge pull request #3 from mozmeao/amend-settings-naming

Amend settings naming and update docs

stevejalim committed:
Small refactor to move the link generation out of the view, to make it more reusable from Python

stevejalim committed:
Bump mozmeao version to 0.1.3

stevejalim committed:
Merge pull request #4 from mozmeao/refactor-link-generation-to-be-a-python-api

Move the link generation out of the view, to make it more reusable from Python

stevejalim committed:
Refactor settings to be a dict

stevejalim committed:
Allow the max age of a sharing link to be set via method arguments, not just via settings

stevejalim committed:
Rename max_age to max_ttl, ahead of adding min_ttl

stevejalim committed:
Also rename MAX_AGE setting to MAX_TTL to match max_tt method argument

stevejalim committed:
Fix erroneous test names

stevejalim commented:
Note: our fork of the project is moving on beyond what the description covers here.

stevejalim committed:
Support making a new sharing link each time, not reusing an existing one

This avoids the downside of reusing links: you might get a link that will expire soon
(or has expired but won't be marked as such until it's used)

stevejalim committed:
Increase default link TTL to 28 days

stevejalim committed:
Add extra audit logging to show when we're reusing an existing link vs generating a new one

stevejalim committed:
Prep 0.2.0 release

stevejalim committed:
Update documentation

stevejalim committed:
Merge pull request #7 from mozmeao/5-customisable-max-age

Support custom max age + multiple links per Revision

mixxorz commented:
Thank you so much! I'll merge this in.
